### DISCUSSION TITLE
LADSPA Reverb/Chorus Send: from mono to stereo

### DISCUSSION THREAD

**Discussion by md1872b** [FluidSynth/fluidsynth#1161] [Q&A] [answered]
Hello, I'm testing some custom LADSPA effects (both Win and Linux). CC91/CC93 work correctly and i'm happy.

However, the reverb and chorus send are mono.
I'm trying to extend this to provide LADSPA fx unit with reverb and chorus stereo send.

So I'd like to obtain something like this for LADSPA binding:
Main:L, MainR
Reverb:SendL, Reverb:SendR
Chorus:SendL, Chorus:SendR

Do you have any tips in exteding this?

I already saw I need to change the buffer for rvoice and adjusting the gain calculation for each buffer
> #define FLUID_RVOICE_MAX_BUFS (6)

**Comment by derselbst** [comment, marked as answer]
@mawe42  is the LADSPA expert, but he seems to be busy.

So, yes the input channels used for reverb and chorus are mono currently. Even for fluidsynth's internal effects. In both cases, the fx unit is expected to create a stereo output signal. You may find more information [here](https://github.com/FluidSynth/fluidsynth/blob/master/doc/ladspa.md#fluidsynth-host-ports) and some thoughts about improving this [here](https://github.com/FluidSynth/fluidsynth/pull/1117#issuecomment-1159810181).

> Do you have any tips in exteding this?
> I already saw I need to change the buffer for rvoice and adjusting the gain calculation for each buffer

You seem to be on the right track. Rather than using an ugly macro with hard-coded indexing for `fluid_rvoice_buffers_set_amp` and friends (as done e.g. [here](https://github.com/FluidSynth/fluidsynth/blob/5dcae73657559f1606d56f0a8b5cde8d53c79ebd/src/synth/fluid_voice.c#L772)), one should think about defining an enum instead

```c
typedef enum  {
DRY_L,
DRY_R,
REV_L,
REV_R,
CHO_L,
CHO_R,
RBUF_COUNT
}  fluid_rvoice_bufidx_t;
```

The interesting part will be to adjust the gain calculation of each buffer, here:

https://github.com/FluidSynth/fluidsynth/blob/5dcae73657559f1606d56f0a8b5cde8d53c79ebd/src/synth/fluid_voice.c#L805-L817

One might think about changing it to something like

```c
// left reverb buffer
fluid_voice_calculate_gain_amplitude(voice, fluid_pan(voice->pan, 1) * fluid_balance(voice->balance, 1) * voice->reverb_send);
```

Similar for the right buffer, just like we have done for the CUSTOM_BALANCE generator just a few lines above.

More gain logic also lives here, not sure if there are other places which I'm missing...

https://github.com/FluidSynth/fluidsynth/blob/5dcae73657559f1606d56f0a8b5cde8d53c79ebd/src/synth/fluid_voice.c#L1848-L1873

**Reply by md1872b** [reply]
Thank you, I was almost close.
Now I'm facing some issues with the extension of fluid_rvoice_mixer in order to support the new buffers.
That's somewhat tricky.

**Comment by md1872b** [comment]
Got it, work as expected.

The issue was I didn't understand that I should multiply by 2 the `fx_buf_count` on the return of `fluid_mixer_buffers_prepare`, something like `return offset + 2 * buffers->fx_buf_count`

**Reply by derselbst** [reply]
Glad to hear! I opened a PR (#1170) and will test it myself once I find the time.
